File: apps/yryvu-app/src/ipc/integrations/oauth.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

/**
 * Result of `oauth_begin` — the URL to open in the user's browser plus
 * an opaque session id that `oauth_await` / `oauth_cancel` need.
 */
export interface OAuthBeginResult {
  sessionId: string;
  authorizeUrl: string;
}

/**
 * Phase 1 of the OAuth flow. The backend binds an ephemeral loopback
 * port, generates PKCE + CSRF state, builds the provider's authorize
 * URL, and parks the session under `sessionId` for `oauth_await` to
 * pick up. The caller is expected to open `authorizeUrl` via
 * `tauri-plugin-opener` and persist `sessionId` until
 * `oauth_await` returns.
 *
 * Errors:
 * - `"OAuth not configured"` — provider's `client_id` is empty (no
 *   `.env.local` / Repo Secret bake). Frontend should disable the
 *   OAuth button when `provider.isConfigured()` is false to avoid
 *   surfacing this.
 */
export function oauthBegin(integrationType: string): Promise<OAuthBeginResult> {
  return invoke<OAuthBeginResult>("oauth_begin", { integrationType });
}

/**
 * Phase 2 of the OAuth flow. Blocks until the provider redirects to
 * the loopback (default 5-minute window in the backend), then exchanges
 * the authorization code for an access token. Returns the raw token —
 * the caller pipes it through `saveToken` (preflight + persist).
 *
 * Errors:
 * - `"OAuth flow cancelled by user"` — user denied consent in the browser.
 * - `"OAuth state mismatch"` — CSRF defense triggered.
 * - `"OAuth flow timed out"` — no redirect arrived in 5 min.
 * - `"OAuth code-for-token exchange failed: <detail>"` — provider rejected the code.
 * - `"OAuth session not found"` — `sessionId` was never `oauth_begin`'d
 *   or has been consumed by a previous await.
 */
export function oauthAwait(sessionId: string): Promise<string> {
  return invoke<string>("oauth_await", { sessionId });
}

/**
 * Drop a parked OAuth session by `sessionId` without consuming it.
 * Releases the bound loopback port. Always succeeds — calling cancel
 * for an unknown id is a no-op.
 */
export function oauthCancel(sessionId: string): Promise<void> {
  return invoke<void>("oauth_cancel", { sessionId });
}

File: crates/yryvu-bridge/src/commands/integrations/oauth.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use serde::Serialize;

use crate::integrations::oauth;

/// Returned by [`oauth_begin`]. The frontend opens `authorize_url` in
/// the user's browser via `tauri-plugin-opener` and persists
/// `session_id` so [`oauth_await`] can pick the session up again.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OAuthBeginResult {
    pub session_id: String,
    pub authorize_url: String,
}

/// Phase 1 of the OAuth flow: build the authorize URL for `integration_type`,
/// stage the [`OAuthSession`] in the registry, and return both the URL
/// and an opaque `session_id` to the frontend.
///
/// [`OAuthSession`]: crate::integrations::oauth::OAuthSession
#[tauri::command]
pub async fn oauth_begin(integration_type: String) -> Result<OAuthBeginResult, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let (session, authorize_url) =
            oauth::flow::begin(&integration_type).map_err(|e| e.to_string())?;
        // 128-bit base64 random — opaque to the frontend, decoupled
        // from the CSRF token that already lives in the URL's `state`.
        let session_id = oauth2::CsrfToken::new_random().secret().to_string();
        oauth::state::insert(session_id.clone(), session);
        Ok(OAuthBeginResult {
            session_id,
            authorize_url,
        })
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Phase 2 of the OAuth flow: take the parked session by `session_id`,
/// block on the loopback redirect (default 5-min window), and exchange
/// the authorization code for an access token. The token is returned
/// raw — the frontend pipes it through `saveToken` (preflight + persist).
#[tauri::command]
pub async fn oauth_await(session_id: String) -> Result<String, String> {
    let session = oauth::state::take(&session_id)
        .ok_or_else(|| "OAuth session not found (already consumed or cancelled)".to_string())?;
    oauth::flow::await_completion(session, oauth::flow::DEFAULT_FLOW_TIMEOUT)
        .await
        .map_err(|e| e.to_string())
}

/// Drop a parked OAuth session by `session_id` without consuming it.
/// Used when the user dismisses the connect dialog before the browser
/// round-trip completes. Releases the bound loopback port.
#[tauri::command]
pub async fn oauth_cancel(session_id: String) -> Result<(), String> {
    oauth::state::drop_session(&session_id);
    Ok(())
}
